### ISSUE TITLE
Sparsifying the data structures in the cell mosaics

### ISSUE THREAD

**Issue by beyondmetis** [isetbio/isetbio#242]
It looks like the mosaics in isetbio are processed using dense rectangular matrices, even though the underlying biological connections are not dense. This scales horribly when we request larger fovs. Specifically, both absorptions and current in the cone mosaic are stored as a dense matrix, which further increases memory storage requirements. With the current design, I don't think processing photocurrent of the whole retina for multiple seconds is possible from a memory standpoint.

To measure the relative amount of sparsity, I computed a hex mosaic centered on the fovea with an area of 9 square mm. Only 7.8% of the storage used in this mosaic is actual cone data.

In Issue #136, @DavidBrainard mentioned sparse matrices, which I think are the way to go. Has anyone worked on this issue?

**Comment by npcottaris** [comment]
I do not think that processing of the entire retina was ever an option that the team considered feasible.
The approach was always that of modeling retinal patches. Beyond the memory footprint issue that you raise, there is the issue of eccentricity-varying cell properties, such as temporal dynamics, signal/noise ratio, off-axis optics, etc, which would significantly complicate coding and which are not well known. The only eccentricity-based cell property that is modeled is cone density. This could benefit from sparse matrices, but no-one has worked on this yet.

**Comment by render-toolbox** [comment]
I concur that we have adopted a patch-based approach.  But note that we do have options for varying more the cell-density with eccentricity (optics, outer segment dynamics, etc.).

It might well go faster with sparse matrices.  I am also aware of a group that used a GPU to compute isomerizations on a pretty big mosaic, but I don't have access to that code.

**Comment by npcottaris** [comment]
Yes, we do have options for setting these, but they cannot be varied continuously with eccentricity across the simulated patch. At the moment, only cell density can vary continuously. 

Nicolas
